### DISCUSSION TITLE
Optimizing Nuxt CSR applications loading time with shared dependencies?

### DISCUSSION THREAD

**Discussion by nicolaspayot** [nuxt/nuxt#28317] [Questions]
At my company we are using different Nuxt 3 applications (SSR and CSR) at different places of our platform/website. One URL can redirect the traffic to a specific Nuxt app while another URL will redirect the traffic to another Nuxt app, and so on.

We're trying to optimize the loading time of the Nuxt CSR applications. At the moment, when a user navigates between URLs that are in different Nuxt applications, we display a loader (`<NuxtLoadingIndicator />`), which makes the UX not great.

I was wondering if there was anything we could do to improve this, such as extracting in a specific bundle some common dependencies between Nuxt applications. Once it's in the user's cache, it could improve the load time. Is this something that could be possible?

We can't use layers (yet) due to organization constraints but that's something we're could be thinking about, to regroup some applications into a bigger one with layers. Also we don't want to convert them to SSR due to cost constraints.

**Comment by ChristianUlbrich** [comment]
What are your shared dependencies? If they are third party libraries (e.g. `lodash`), you could definitely _externalize_ them and serve them from a CDN. That way they are only loaded and parsed once and are available for all apps. This does not seem to be trivial within Nuxt, as it would have to involve configure the bundling process, best bet is probably using [dynamic imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) from the CDN.
If they are shared Vue components, you could also have them externalized by using [Async Components](https://v3.ru.vuejs.org/guide/component-dynamic-async.html) (and use also _dynamic imports_).

If you externalize any of those, those become _dependencies_ of your various apps and you need to make sure, that your apps are working with the same version, i.e. if you have to update a third party library for App A, that you are also using in App B, you have to make sure, that App B is still working. Usually this approach works best with a _mono repo_, but then you have to figure out individual releases.
